cdap-app-fabric/src/main/java/co/cask/cdap/metadata/DefaultEntityExistenceVerifier.java

```diff
@@ -0,0 +1,197 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.metadata;
+
+import co.cask.cdap.api.app.ApplicationSpecification;
+import co.cask.cdap.api.dataset.DatasetManagementException;
+import co.cask.cdap.app.store.Store;
+import co.cask.cdap.common.ApplicationNotFoundException;
+import co.cask.cdap.common.ArtifactNotFoundException;
+import co.cask.cdap.common.DatasetNotFoundException;
+import co.cask.cdap.common.NamespaceNotFoundException;
+import co.cask.cdap.common.NotFoundException;
+import co.cask.cdap.common.ProgramNotFoundException;
+import co.cask.cdap.common.StreamNotFoundException;
+import co.cask.cdap.common.ViewNotFoundException;
+import co.cask.cdap.common.entity.EntityExistenceVerifier;
+import co.cask.cdap.data2.dataset2.DatasetFramework;
+import co.cask.cdap.data2.transaction.stream.StreamAdmin;
+import co.cask.cdap.internal.app.runtime.artifact.ArtifactStore;
+import co.cask.cdap.proto.ProgramType;
+import co.cask.cdap.proto.id.ApplicationId;
+import co.cask.cdap.proto.id.DatasetId;
+import co.cask.cdap.proto.id.EntityId;
+import co.cask.cdap.proto.id.NamespaceId;
+import co.cask.cdap.proto.id.NamespacedArtifactId;
+import co.cask.cdap.proto.id.ProgramId;
+import co.cask.cdap.proto.id.ProgramRunId;
+import co.cask.cdap.proto.id.StreamId;
+import co.cask.cdap.proto.id.StreamViewId;
+import co.cask.cdap.store.NamespaceStore;
+import com.google.common.base.Preconditions;
+import com.google.common.base.Throwables;
+import com.google.inject.Inject;
+
+import java.io.IOException;
+import java.util.Set;
+
+/**
+ * Default implementation of {@link EntityExistenceVerifier}.
+ */
+public class DefaultEntityExistenceVerifier implements EntityExistenceVerifier {
+  private final NamespaceStore nsStore;
+  private final ArtifactStore artifactStore;
+  private final Store store;
+  private final DatasetFramework datasetFramework;
+  private final StreamAdmin streamAdmin;
+
+  @Inject
+  DefaultEntityExistenceVerifier(NamespaceStore nsStore, ArtifactStore artifactStore, Store store,
+                                 DatasetFramework datasetFramework, StreamAdmin streamAdmin) {
+    this.nsStore = nsStore;
+    this.artifactStore = artifactStore;
+    this.store = store;
+    this.datasetFramework = datasetFramework;
+    this.streamAdmin = streamAdmin;
+  }
+
+  @Override
+  public void ensureExists(EntityId entityId) throws NotFoundException {
+    if (entityId instanceof NamespaceId) {
```

Comment:
Instead of doing this `instanceof`, isn't that you can just use method overload?

```java
public void ensureExists(NamespaceId id);
public void ensureExists(NamespacedArtifactId id);
public void ensureExists(ApplicationId id);
...
```

I think it's cleaner than doing the dispatching manually.